### PR TITLE
Integrate wrestic into k8up

### PR DESCRIPTION
**Pull request by cimnine**
## Summary

Integrates wrestic into the k8up repository and unifies them into a single binary.
This is the bulk-load of changes required to finish https://github.com/vshn/k8up/issues/359 eventually.

Tasklist:

- [x] Single CLI interface
- [x] e2e test using wrestic shim
- [x] integration test
- [x] ~unification of configuration methods~ See these instead:
  * #477
  * #481
- [x] planning the removal of the wrestic shim
  * #476 
- [x] ~docs~  See these instead:
  * #482

## Checklist

- [x] ~Keep pull requests small so they can be easily reviewed.~ _I tried my best, but that's really hard here._
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`,
      as they show up in the changelog
- [x] Update the documentation.
- [x] Update tests.
- [x] Link this PR to related issues.

### CODE REVIEW

**Review by ccremer — COMMENTED** [review]
First look at the PR.
I didn't look into the restic/cli/* files to deeply, I imagine they are mostly 1:1 from wrestic repo.

File: cmd/k8up/main.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+		EnableBashCompletion: true,
+		Before:               mainAction,
+		Flags: []cli.Flag{
+			&cli.BoolFlag{
+				Name:        "debug",
+				Aliases:     []string{"verbose", "d"},
+				Usage:       "sets the log level to debug",
+				EnvVars:     []string{"K8UP_DEBUG"},
+				DefaultText: "false",
+			},
```
[the remark this answers is not in the document]
Comment (cimnine) [reply cmd/k8up/main.go:62]: `-v` ist schon durch `version` besetzt.

File: cmd/restic/main.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+	ctx, cancel := context.WithCancel(c.Context)
+	cancelOnTermination(cancel, resticLog)
+
+	statHandler := stats.NewHandler(os.Getenv(promURLEnv), os.Getenv(resticCli.Hostname), os.Getenv(webhookURLEnv), resticLog)
+
+	resticCLI := resticCli.New(ctx, resticLog, statHandler)
+
+	return run(c.Context, resticCLI, resticLog)
+}
```
[the remark this answers is not in the document]
Comment (cimnine) [reply cmd/restic/main.go:75]: I have opened a new issues for this, issue #481.

File: docs/modules/ROOT/pages/explanations/architecture.adoc [hunk 3]
```diff
@@ -3,7 +3,7 @@
 K8up consists of two parts:
 
 * Kubernetes Operator: https://github.com/vshn/k8up[K8up]
-* Restic wrapper: https://github.com/vshn/wrestic/[wrestic]
+* Restic wrapper: https://github.com/vshn/k8up/[wrestic]
```
Comment (simu) [inline docs/modules/ROOT/pages/explanations/architecture.adoc:6]: Maybe link the wrestic package in K8up here?

**Review by simu — APPROVED** [review]
LGTM overall, but I'm not familiar enough with either code base to confidently say whether everything got reorganized correctly.

File: .goreleaser.yml [hunk 4]
```diff
@@ -16,10 +18,9 @@ snapshot:
   name_template: "{{ .Tag }}-snapshot"
 
 dockers:
-
 - image_templates:
-  - "docker.io/vshn/k8up:v{{ .Version }}"
-  - "quay.io/vshn/k8up:v{{ .Version }}"
+  - "docker.io/vshn/k8up:{{ .Version }}"
+  - "quay.io/vshn/k8up:{{ .Version }}"
```
Comment (ccremer) [inline .goreleaser.yml:23]: The `.Version` comes without the `v` prefix, so we should keep it there.
https://goreleaser.com/customization/templates/

File: .goreleaser.yml [hunk 5]
```diff
@@ -29,6 +30,42 @@ dockers:
...
+  - "quay.io/vshn/wrestic:{{ .Version }}-nonroot"
+
+  # For prereleases, updating `latest` and the floating tags of the major
+  # version does not make sense. Only the image for the exact version should
+  # be pushed.
+  - "{{ if not .Prerelease }}docker.io/vshn/wrestic:v{{ .Major }}{{ end }}-nonroot"
+  - "{{ if not .Prerelease }}quay.io/vshn/wrestic:v{{ .Major }}{{ end }}-nonroot"
+
+  - "{{ if not .Prerelease }}docker.io/vshn/wrestic:latest{{ end }}-nonroot"
+  - "{{ if not .Prerelease }}quay.io/vshn/wrestic:latest{{ end }}-nonroot"
```
Comment (ccremer) [inline .goreleaser.yml:48]:
```suggestion
  - "{{ if not .Prerelease }}docker.io/vshn/wrestic:v{{ .Major }}-nonroot{{ end }}"
  - "{{ if not .Prerelease }}quay.io/vshn/wrestic:v{{ .Major }}-nonroot{{ end }}"

  - "{{ if not .Prerelease }}docker.io/vshn/wrestic:latest-nonroot{{ end }}"
  - "{{ if not .Prerelease }}quay.io/vshn/wrestic:latest-nonroot{{ end }}"
```

File: Makefile [hunk 6]
```diff
@@ -6,48 +6,56 @@ MAKEFLAGS += --no-builtin-rules
 MAKEFLAGS += --no-builtin-variables
 .SUFFIXES:
 .SECONDARY:
+.DEFAULT_GOAL := all
```
Comment (ccremer) [inline Makefile:9]:
```suggestion
.DEFAULT_GOAL := help
```

**Review by ccremer — APPROVED** [review]
just some config errors, rest LGTM

File: .goreleaser.yml (hunk 4 above)
Comment (cimnine) [reply .goreleaser.yml:23]: Interesting. If I run goreleaser on my host, this is what I get:

```
   • docker images
      • building docker image     image=docker.io/vshn/wrestic:v1.2.0-snapshot
      • building docker image     image=docker.io/vshn/k8up:v1.2.0-snapshot
      • building docker image     image=docker.io/vshn/wrestic:v1.2.0-snapshot-nonroot
      • pipe skipped              error=publishing is disabled
```

But if I change it to `"docker.io/vshn/k8up:{{ .Version }}"`, then I get:

```
   • docker images
      • building docker image     image=docker.io/vshn/k8up:vv1.2.0-snapshot
      • building docker image     image=docker.io/vshn/wrestic:vv1.2.0-snapshot
      • building docker image     image=docker.io/vshn/wrestic:vv1.2.0-snapshot-nonroot
      • pipe skipped              error=publishing is disabled
```

That's why I dropped the 'v' in the first place.

Comment (cimnine) [reply .goreleaser.yml:23]: I've found the bug. I was running `goreleaser --snapshot` and in this mode the substitution of `vX.Y.Z` to `X.Y.Z` does not seem to work.

File: cmd/restic/args/args.go [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package args
```
Comment (Kidswiss) [inline cmd/restic/args/args.go:1]: Can't `urfave/cli/v2` handle this?

**Review by Kidswiss — APPROVED** [review]
Just had one small remark. That could very well also be handled in another issue/MR.

File: cmd/restic/args/args.go (hunk 7 above)
Comment (cimnine) [reply cmd/restic/args/args.go:1]: These are the arguments to pass to restic, not the args retrieved from the CLI. That's why I don't see `urface/cli/v2` for that purpose.

Comment (cimnine) [reply cmd/restic/args/args.go:1]: You're more right than I was and it turns out this is just dead code.

**Pull request merged**